app/src/main/java/org/dhis2/utils/customviews/PeriodAdapter.kt

```diff
@@ -70,5 +79,38 @@ private class PeriodAdapter(
         }
     }
 
-    override fun getItemCount() = datePeriods.size + 1
+    fun setInputPeriod(){
+        var isAllowed = false
+        for (inputPeriodModel in inputPeriods) {
+            do {
+                if ((currentDate.after(inputPeriodModel.initialPeriodDate()) || currentDate == inputPeriodModel.initialPeriodDate()) && currentDate.before(
+                        inputPeriodModel.endPeriodDate()
+                    )
+                    && (inputPeriodModel.openingDate() == null || inputPeriodModel.openingDate() != null && DateUtils.getInstance().today.after(
+                        inputPeriodModel.openingDate()
+                    )
+                            || DateUtils.getInstance().today == inputPeriodModel.openingDate())
+                    && (inputPeriodModel.closingDate() == null || inputPeriodModel.closingDate() != null && DateUtils.getInstance().today.before(
+                        inputPeriodModel.closingDate()
+                    ))
+                ) {
+                    isAllowed = true
+                    datePeriods.add(currentDate)
+                    lastDate = currentDate
+                }
+                else if (currentDate.before(inputPeriods[inputPeriods.size - 1].initialPeriodDate()) || currentDate.before(
+                        inputPeriodModel.initialPeriodDate()
+                    )
+                )
+                    break
+                else {
+                    currentDate = DateUtils.getInstance().getNextPeriod(periodType, currentDate, -1)
+                }
+            } while (!isAllowed)
```

Comment:
I think is eassier to follow this code if you change the "do while" to a "for" loop :) 

Please when adding so much logic its better to be in classes that this can be testest, this could have different corner cases that could be identified during unit tests. We need to avoid adding logic to android components :sob: